File: solutions/2026-08-04/javascript/2026-08-04-find-missing-elements.js
/*
Problem Summary:
This problem requires finding all missing integers within a continuous range, given an array `nums` where the smallest and largest integers of that original range are guaranteed to be present. The output should be a sorted list of these missing integers.

Link:
https://leetcode.com/problems/find-missing-elements/

Approach Explanation:
1.  **Determine the Range**: First, identify the minimum and maximum values present in the input array `nums`. These values define the complete range of integers that should have been in the original set.
2.  **Efficient Lookup**: To quickly check for the presence of numbers, convert the input array `nums` into a `Set`. This allows for average O(1) time complexity for checking if a number exists.
3.  **Iterate and Collect**: Iterate through every integer from the determined minimum value up to the maximum value (inclusive). For each integer in this sequence, check if it is present in the `Set`. If an integer is not found in the `Set`, it means it is a missing element, so add it to a result list.
4.  **Return Sorted List**: Since the iteration happens in ascending order from minimum to maximum, the collected missing elements will naturally be in sorted order. Return this list.

Time Complexity:
-   Finding min and max values: O(N), where N is the length of `nums`.
-   Creating the `Set` from `nums`: O(N), as each insertion into a Set takes average O(1) time.
-   Iterating from `minVal` to `maxVal` and checking the Set: O(M), where M is the size of the range (`maxVal - minVal + 1`). Each Set lookup takes average O(1) time.
-   Overall Time Complexity: O(N + M). Given the constraints (N <= 100, `nums[i]` <= 100, so M <= 100), this is very efficient.

Space Complexity:
-   `numSet`: O(N) space to store all unique elements from `nums`.
-   `missingNumbers`: O(M) space in the worst case, where M is the range size (e.g., if `nums = [1, 100]`, almost all numbers are missing).
-   Overall Space Complexity: O(N + M).
*/
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findMissingElements = function(nums) {
    // Determine the minimum and maximum values in the nums array.
    // These define the bounds of the original integer range.
    let minVal = Infinity;
    let maxVal = -Infinity;
    for (const num of nums) {
        if (num < minVal) {
            minVal = num;
        }
        if (num > maxVal) {
            maxVal = num;
        }
    }

    // Create a Set from the input array for efficient O(1) average time lookups.
    const numSet = new Set(nums);

    // Initialize an array to store the missing numbers.
    const missingNumbers = [];

    // Iterate through the entire range from minVal to maxVal (inclusive).
    for (let i = minVal; i <= maxVal; i++) {
        // If the current number 'i' is not found in our Set, it means it's missing.
        if (!numSet.has(i)) {
            // Add the missing number to our result list.
            // Since we iterate in ascending order, this list will naturally be sorted.
            missingNumbers.push(i);
        }
    }

    // Return the list of all missing numbers.
    return missingNumbers;
};

File: solutions/2026-08-04/python/2026-08-04-find-missing-elements.py
# Problem Summary:
# Given an array of unique integers, find all integers missing from the range defined by the minimum and maximum elements present in the array.
# The original array is guaranteed to contain the smallest and largest integers of its full range.
# Link: https://leetcode.com/problems/find-missing-elements/

# Approach Explanation:
# 1. Determine the full range: First, find the minimum and maximum values within the input `nums` array. These two values define the inclusive boundaries of the complete sequence of numbers that should have been present.
# 2. Create an efficient lookup structure: Convert the input `nums` list into a hash set (Python's `set`). This data structure provides average O(1) time complexity for checking if an element exists, which is crucial for efficient lookups.
# 3. Iterate through the full range: Generate all integers from the calculated minimum value up to the maximum value (inclusive).
# 4. Identify missing elements: For each integer in this generated range, check if it is present in the hash set created from `nums`. If an integer is not found in the set, it means it was missing from the original array.
# 5. Collect and return: Add all identified missing integers to a list. Since we iterate through the range in ascending order, the resulting list of missing integers will naturally be sorted. If no integers are missing, the list will remain empty.

# Time Complexity:
# O(N + R) where N is the number of elements in `nums` and R is the size of the determined range (max_val - min_val + 1).
# - Finding `min_val` and `max_val`: O(N) using built-in `min()` and `max()`.
# - Creating the `nums_set`: O(N) to add all N elements to the set.
# - Iterating from `min_val` to `max_val` (R elements) and performing set lookups (O(1) average for each): O(R).
# Given the constraints (N <= 100, `nums[i]` <= 100), R will also be at most 100. Thus, the total time complexity is very efficient.

# Space Complexity:
# O(N + R) where N is the number of elements in `nums` and R is the size of the determined range.
# - `nums_set`: O(N) space to store the unique elements from `nums`.
# - `missing_elements` list: In the worst-case scenario (e.g., `nums = [1, 100]` and all numbers between 1 and 100 except 1 and 100 are missing), this list could store up to O(R) elements.
# The total space complexity is dominated by the storage needed for the set and the result list.

class Solution:
    def findMissingElements(self, nums: list[int]) -> list[int]:
        # Step 1: Find the smallest and largest integers in the given array.
        # These values define the inclusive boundaries of the expected full range.
        min_val = min(nums)
        max_val = max(nums)

        # Step 2: Convert the input list `nums` into a set for efficient O(1) average time lookups.
        # This allows us to quickly check if a number from the full range is present in `nums`.
        nums_set = set(nums)

        # Step 3: Initialize an empty list to store the missing integers.
        missing_elements = []

        # Step 4: Iterate through every integer from `min_val` to `max_val` (inclusive).
        # This covers the entire expected range of numbers.
        for i in range(min_val, max_val + 1):
            # Step 5: For each number `i` in the range, check if it is present in `nums_set`.
            # If `i` is not in `nums_set`, it means it's a missing element from the original array.
            if i not in nums_set:
                # Step 6: Add the missing element to our `missing_elements` list.
                # Since we iterate through the range in ascending order, the list will automatically be sorted.
                missing_elements.append(i)
        
        # Step 7: Return the list of all found missing elements.
        # If no elements were missing, this list will be empty.
        return missing_elements